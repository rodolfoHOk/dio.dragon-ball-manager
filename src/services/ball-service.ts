import { Ball } from '../models/ball';

export class BallService {
  private static balls: Ball[] = [
    {
      id: 1,
      name: '1 estrela',
      owner: 0,
      image: '/images/dragonball-1-star.png',
    },
    {
      id: 2,
      name: '2 estrelas',
      owner: 0,
      image: '/images/dragonball-2-stars.png',
    },
    {
      id: 3,
      name: '3 estrelas',
      owner: 0,
      image: '/images/dragonball-3-stars.png',
    },
    {
      id: 4,
      name: '4 estrelas',
      owner: 1,
      image: '/images/dragonball-4-stars.png',
    },
    {
      id: 5,
      name: '5 estrelas',
      owner: 0,
      image: '/images/dragonball-5-stars.png',
    },
    {
      id: 6,
      name: '6 estrelas',
      owner: 0,
      image: '/images/dragonball-6-stars.png',
    },
    {
      id: 7,
      name: '7 estrelas',
      owner: 0,
      image: '/images/dragonball-7-stars.png',
    },
  ];

  static getBalls(): Ball[] {
    return this.balls;
  }

  static foundedBall(ownerId: number, ballId: number) {
    this.balls = this.balls.map((ball) => {
      if (ball.id === ballId) {
        ball.owner = ownerId;
      }
      return ball;
    });
  }

  static hasAllBalls(ownerId: number) {
    const balls = this.balls.filter((ball) => ball.owner === ownerId);
    return balls.length === 7;
  }
}
