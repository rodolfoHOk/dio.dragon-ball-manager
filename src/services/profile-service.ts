import { Profile } from '../models/profile';

export class ProfileService {
  private static profiles: Profile[] = [
    {
      id: 1,
      name: 'Rudy HiOk',
      image:
        'https://raw.githubusercontent.com/rodolfoHOk/portfolio-img/main/images/avatar/pedro.jpg',
      balls: [1, 2, 4, 6, 7],
    },
    {
      id: 2,
      name: 'Jao Pones',
      image:
        'https://raw.githubusercontent.com/rodolfoHOk/portfolio-img/main/images/avatar/joao.jpg',
      balls: [3, 5],
    },
  ];

  static getProfile(id: number): Profile | undefined {
    return this.profiles.find((profile) => profile.id === id);
  }

  static addBall(profileId: number, ballId: number) {
    this.profiles = this.profiles.map((profile) => {
      if (profile.id === profileId) {
        profile.balls.push(ballId);
      } else {
        if (profile.balls.includes(ballId)) {
          profile.balls = profile.balls.filter((ball) => ball !== ballId);
        }
      }
      return profile;
    });
  }
}
