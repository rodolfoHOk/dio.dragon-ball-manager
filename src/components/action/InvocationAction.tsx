import { useState } from 'react';
import { motion } from 'framer-motion';
import { Button } from '../button/Button';
import { AppToast } from '../app-toast/AppToast';
import { BallService } from '../../services/ball-service';

type InvocationActionProps = {
  profileId: number;
};

export function InvocationAction({ profileId }: InvocationActionProps) {
  const [showToast, setShowToast] = useState(false);
  const [showShenlong, setShowShenlong] = useState(false);

  function invocate() {
    const hasAllBalls = BallService.hasAllBalls(profileId);
    if (!hasAllBalls) {
      setShowToast(true);
      return;
    }
    setShowShenlong(true);
    setTimeout(() => setShowShenlong(false), 6000);
  }

  function onShowToast(value: boolean) {
    setShowToast(value);
  }

  return (
    <>
      {showShenlong && (
        <motion.div
          className="absolute top-0 left-0 h-screen w-screen bg-black"
          initial={{ opacity: 0 }}
          animate={{ opacity: 0.97 }}
          transition={{ duration: 1 }}
        >
          <motion.div
            className="h-[80vh] mt-[10vh] flex items-center justify-center"
            initial={{ scale: 0, opacity: 0 }}
            animate={{ scale: 1, opacity: 1 }}
            transition={{ delay: 1.5, duration: 2 }}
          >
            <img
              data-testid="shenlong-image"
              src="/images/shenlong.webp"
              alt="Shenlong apareceu"
            />
          </motion.div>
        </motion.div>
      )}

      <div
        data-testid="invocation-card"
        className="p-6 flex flex-col items-center justify-center gap-4 bg-zinc-100 dark:bg-zinc-900"
      >
        <h2 className="font-bold text-2xl">Invocar Shenlong</h2>

        <Button data-testid="invoke-button" className="w-48" onClick={invocate}>
          Invocar
        </Button>
      </div>

      <AppToast
        message="Você não tem todas as esferas para invocar o Shenlong"
        showToast={showToast}
        setShowToast={onShowToast}
      />
    </>
  );
}
