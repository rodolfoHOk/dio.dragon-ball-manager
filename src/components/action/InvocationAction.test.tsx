import { render, screen, fireEvent } from '@testing-library/react';
import { InvocationAction } from './InvocationAction';
import { ToastProvider, ToastViewport } from '@radix-ui/react-toast';
import { BallService } from '../../services/ball-service';

describe('invocation action component test', () => {
  it('should render invocation action', () => {
    const { container } = render(
      <ToastProvider>
        <div>
          <InvocationAction profileId={1} />
          <ToastViewport />
        </div>
      </ToastProvider>
    );

    expect(screen.getByText('Invocar Shenlong')).toBeInTheDocument();
    expect(screen.getByText('Invocar')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it('should show toast if user do not have all balls', () => {
    render(
      <ToastProvider>
        <div>
          <InvocationAction profileId={1} />
          <ToastViewport />
        </div>
      </ToastProvider>
    );

    const button = screen.getByText('Invocar');
    fireEvent.click(button);
    expect(
      screen.getByText('Você não tem todas as esferas para invocar o Shenlong')
    ).toBeInTheDocument();

    const backButton = screen.getByText('Voltar');
    fireEvent.click(backButton);
    expect(
      screen.queryByText(
        'Você não tem todas as esferas para invocar o Shenlong'
      )
    ).toBeNull();
  });

  it('should show shenlong', () => {
    render(
      <ToastProvider>
        <div>
          <InvocationAction profileId={1} />
          <ToastViewport />
        </div>
      </ToastProvider>
    );

    jest.spyOn(BallService, 'hasAllBalls').mockImplementation(() => true);

    const button = screen.getByText('Invocar');
    fireEvent.click(button);
    expect(screen.getByAltText('Shenlong apareceu')).toBeInTheDocument();
    expect(
      screen.queryByText(
        'Você não tem todas as esferas para invocar o Shenlong'
      )
    ).toBeNull();
  });
});
