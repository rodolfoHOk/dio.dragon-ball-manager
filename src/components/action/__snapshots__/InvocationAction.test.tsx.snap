// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invocation action component test should render invocation action 1`] = `
<div>
  <div>
    <div
      class="p-6 flex flex-col items-center justify-center gap-4 bg-zinc-100 dark:bg-zinc-900"
      data-testid="invocation-card"
    >
      <h2
        class="font-bold text-2xl"
      >
        Invocar Shenlong
      </h2>
      <button
        class="h-10 px-4 flex flex-row items-center justify-center font-bold rounded  transition-colors duration-300 bg-orange-400 text-red-800 hover:bg-orange-300 w-48"
        data-testid="invocation-button"
      >
        Invocar
      </button>
    </div>
    <div
      aria-label="Notifications (F8)"
      role="region"
      style="pointer-events: none;"
      tabindex="-1"
    >
      <ol
        tabindex="-1"
      />
    </div>
  </div>
</div>
`;
